contracts/Neumark.sol

```diff
@@ -100,18 +99,21 @@ contract Neumark is
     function burnNeumark(uint256 neumarkUlps)
         public
         only(ROLE_NEUMARK_BURNER)
-        acceptAgreement(msg.sender)
         returns (uint256)
     {
-        address owner = msg.sender;
-        uint256 euroUlps = incrementalInverse(_totalEuroUlps, neumarkUlps);
-
-        _totalEuroUlps -= euroUlps;
-
-        assert(mDestroyTokens(owner, neumarkUlps));
+        return burnNeumarkFor(msg.sender, neumarkUlps);
+    }
 
-        LogNeumarksBurned(owner, euroUlps, neumarkUlps);
-        return euroUlps;
+    function burnAllNeumark()
```

Comment:
If we go for the mapping aproach, instead of a boolean we can use an Enum at no extra cost. We can then have states `None, Implicit, Explicit, Cancelled` with the obvious meaning and transitions. We can make transactions fail that would otherwise implicitly or explicitly accept if the `signer` has state `Cancelled`. This prevents other users from sending NMK to an opted-out account. The user can re-activate by explicitly accepting the NMK agreement.

Reply:
I do not think cancellation is as complicated. You cannot opt out from transactions you did/signed. You can only say 'i'm not interested in neumarks that someone sent me' or 'I'm not interested in current balance of my neumarks'
So it is OK that when LockedAccount burns your Neumarks, no one cancels or opts out. Unlock does not opt you out from LockedAccount as well - it is part of the terms.
Voting and Disbursal will have separate agreements attached with all the provisions for people that now have 0 balance of Neumarks but have past fees to claim. You should not need NTHA to claim dividend if concerns are properly separated.
I however agree that implementing cancellation is tricky and if there is a possibility to drop it we should just do it (after talking with Jacek - we should add him to github tomorrow)
The other way to do that is to not to sign agreements for `to` in the `transfer`. You can have balance > 0 but until you do not allow/transfer etc. or use it in any other way you are not a party to this agreement. And we do not need to implement cancel. Let's talk to lawyers.

Reply:
Alternative approach that was discussed with Jacek:
1. There is no implicit signing of agreement. There must be a transaction signed by user to get into agreement (even if it is a chain of internal calls of which one signs original msg.sender up - like in case of LockedAccount). So for example in
```
function mOnTransfer(
        address from,
        address to,
        uint256 // amount
    )
        internal
        acceptAgreement(from)
        acceptAgreement(to)
        returns (bool allow)
    {
        return _transferEnabled;
    }
```
we would remove `acceptAgreement(to)`
2.  Opt-out/cancellation is not necessary and we may remove it
3. In that case, however, we must change how Commitment distributes Neumarks. Currently we issue Neumarks to Commitment smart contract and from that Neumarks are distributed to token holders using `transfer`. This will not sign them up (see above). Possible solutions: 
* change
```
function issueForEuro(uint256 euroUlps)
        public
        only(ROLE_NEUMARK_ISSUER)
        acceptAgreement(msg.sender)
```
into
```
function issueForEuroTo(address to, uint256 euroUlps)
        public
        only(ROLE_NEUMARK_ISSUER)
        acceptAgreement(to)
```
and refactor Commitment contract code that distributes Neumarks. I think dealing with reserved slots will be particularly hard.
* add new function to Neumarks contract that distributes them like
```
function distributeNeumark(address to, uint256 neumarkUlps)
        public
        only(ROLE_NEUMARK_ISSUER)
        acceptAgreement(to)
    {
        bool success = transfer(to, neumarkUlps);
        require(success);
    }
```
and use it instead of `transfer` to distribute Neumark in Commitment. It already has required ROLE_NEUMARK_ISSUER role. This would split `issueForEuroTo` proposed earlier into two parts: allocate to issuer, distribute from issuer

Reply:
1. In this case,  can use `tx.origin` in the `acceptAgreement` modifier and remove the argument? It makes sense because `tx.origin` is the signer of the transaction.

2. Cool. :)

3. Not necessary if we go the `tx.origin` route. The call to `transfer` would make `tx.origin` (the investor) sign the Neumark agreement. This makes sense because by using the Commit contract, you also explicitly agree to the Neumark contract. (and we have your signature). 

I don't like the proposed changes to the Neumark contract functions. It adds complexity and does not guarantee that `acceptAgreement` is called correctly. You need to trust `ROLE_NEUMARK_ISSUE` for this. If we go the `tx.origin` route, no trust is required and the code would be getting even simpler than it is now.

Reply:
ad 1. What about multisig wallets (may be used by organizations to make decisions) or various types of watched wallets like ours or parity one? In case of LockedAccount, signer of the Agreement should be the address on which balance is stored. Exactly the same situation with Neumark token.
In other words `msg.sender` calling `Commitment` contract should be signer both for LockedAccount and Neumark agreements (btw. reading Jacek's acquisition terms, it is pretty obvious we must move Agreement from LockedAccount to Commitment contracts that you write).

Any opinion on that? We should first agree on this before making further changes

Reply:
In case of multisig and other wallets, you would get the signature of the wallet operator that caused the transaction to execute.

There are basically two entities, the direct initiator of the transaction, which will often be a smart contract from us or elsewhere. And there is the ultimate initiation of the transaction, which is  an entity external to the blockchain identified by a private key.

The current proposal tries to create some new kind of entity that is somewhere in between, by forwarding addresses and creating trust relationships. How is this third entity defined? Is it the first non-Neufund address going up the call chain? Does this make sense in the context of an exchange contract?

Reply:
Current proposal says that address holding tokens or funds is party to the agreement. This address is `msg.sender` entering our system (whatever this represents). What is not clear here?